### PR TITLE
Jars Home 4 jrubyExec

### PR DESCRIPTION
**Pull request by mkristian**

### CODE REVIEW

File: jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/GemUtils.groovy [hunk 1]
```diff
@@ -156,6 +156,70 @@ class GemUtils {
...
+        def files = []
+        artifacts.each { dep ->
+
+            def group = dep.moduleVersion.id.group
+            def groupAsPath = group.replace('.' as char, File.separatorChar)
+            def version = dep.moduleVersion.id.version
+            // TODO classifier
+            def newFileName = "${groupAsPath}/${dep.name}/${version}/${dep.name}-${version}.${dep.type}"
+
+            if (group != 'rubygems' && dep.type != 'gem' && dep.name != 'jruby-complete') {
```
Comment (rtyler) [inline jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/GemUtils.groovy]: This conditional could use some logic explaining these special cases that we're ignoring for coordinate computation

File: jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/GemUtils.groovy [hunk 2]
```diff
@@ -156,6 +156,70 @@ class GemUtils {
...
+                // TODO classifier and system-scope
+                coordinates << "${group}:${dep.name}:${version}:runtime:"
+            }
+            fileRenameMap[dep.file.name] = newFileName
+            // TODO omit system-scoped files
+            files << dep.file
+        }
+
+        // create Jars.lock file used by jar-dependencies
+        File jarsLock = new File(destDir, 'Jars.lock')
```
Comment (rtyler) [inline jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/GemUtils.groovy]: IMO L185-L200 should be broken out into its own method which would be easier to write unit tests for, e.g.

```groovy
/**
 * @return boolean true if the Jars.lock file has been successfully written
 * @throws DuplicateFileCopyingException if there is an existing Jars.lock file
 */
static boolean writeJarLockFile(List<String> dependencyCoordinates, GemUtils.OverrideAction overwrite)
```


Then the unit tests for the method can simply verify those three behaviors in isolation

File: jruby-gradle-base-plugin/src/integTest/groovy/com/github/jrubygradle/JRubyExecIntegrationSpec.groovy [hunk 3]
```diff
@@ -83,14 +83,18 @@ class JRubyExecIntegrationSpec extends Specification {
 
     def "Running a script that requires a gem, a separate jRuby and a separate configuration"() {
         given:
+            final String newVersion = '1.7.11'
+            assert project.jruby.execVersion != newVersion
```
Comment (rtyler) [inline jruby-gradle-base-plugin/src/integTest/groovy/com/github/jrubygradle/JRubyExecIntegrationSpec.groovy]: I don't understand why this assertion is here

File: jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/GemUtils.groovy [hunk 4]
```diff
@@ -156,6 +156,70 @@ class GemUtils {
...
+                  break
+                }
+            case OverwriteAction.OVERWRITE:
+                jarsLock.parentFile.mkdirs()
+                jarsLock.withWriter { writer ->
+                    coordinates.each { writer.println it }
+                }
+        }
+
+        files.each { File f ->
```
Comment (rtyler) [inline jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/GemUtils.groovy]: Similar to my comment on the above block, I think this block (L202-L220) can/should be broken out into an independent method so it can be better tested, e.g.

```groovy
/**
 * @return boolean true if we could successfully rewrite the dependency's file
 * @throws DuplicateFileCopyingException failure to overwrite a dependency file which already exists
 */
static boolean rewriteDependencyFiles(List<File> dependencies, Map<String, String> renameMap)
```

The unit test for this method I think we be similarly easy to verify that our `overwrite` behavior is correct

File: jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/JRubyExec.groovy [hunk 5]
```diff
@@ -19,19 +19,25 @@ import org.gradle.util.CollectionUtils
 class JRubyExec extends JavaExec implements JRubyExecTraits {
 
     static final String JRUBYEXEC_CONFIG = 'jrubyExec'
+    static final String JAR_DEPENDENCIES_VERSION = '0.1.15'
 
     static void updateJRubyDependencies(Project proj) {
         proj.dependencies {
             jrubyExec "org.jruby:jruby-complete:${proj.jruby.execVersion}"
-            jrubyExec "org.bouncycastle:bcprov-jdk15on:${proj.jruby.bouncycastleVersion}"
+            if (proj.jruby.execVersion.startsWith("1.7.1")) {
```
Comment (rtyler) [inline jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/JRubyExec.groovy]: Will you link to a changelog entry or something explaining why we do not need to inject a dependency on versions of JRuby >= `1.7.20`

File: jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/JRubyExec.groovy [hunk 6]
```diff
@@ -199,7 +210,11 @@ class JRubyExec extends JavaExec implements JRubyExecTraits {
      */
     @Override
     List<String> getArgs() {
-        JRubyExecUtils.buildArgs(jrubyArgs, getScript(), scriptArgs)
+        List<String> extra = []
+        if (JRUBYEXEC_CONFIG != jrubyConfigurationName) {
```
Comment (rtyler) [inline jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/JRubyExec.groovy]: These needs some comments on why it is here for posterity (read: so I don't accidentally delete it in the future :P)

File: jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/internal/JRubyExecDelegate.groovy [hunk 7]
```diff
@@ -95,13 +96,19 @@ class JRubyExecDelegate implements JRubyExecTraits   {
                 "${k}" v
             }
             main 'org.jruby.Main'
+            // just keep this even if it does not exists
+            args '-I' + new File(gemDir, "gems/jar-dependencies-${JRubyExec.JAR_DEPENDENCIES_VERSION}/lib").absolutePath
```
Comment (rtyler) [inline jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/internal/JRubyExecDelegate.groovy]: This line looks the same as [this line](https://github.com/jruby-gradle/jruby-gradle-plugin/pull/134/files#diff-017bc85ec183b742e079cba05ac40f66R215).

Will you please refactor this behavior into a method which both locations can call?

File: jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/internal/JRubyExecUtils.groovy [hunk 8]
```diff
@@ -79,7 +79,7 @@ class JRubyExecUtils {
 
     static List<String> buildArgs(List<Object> jrubyArgs, File script, List<Object> scriptArgs) {
         def cmdArgs = []
-
+        cmdArgs.add('-rjars/setup')
```
Comment (rtyler) [inline jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/internal/JRubyExecUtils.groovy]: This needs a comment on why it is here/what it is for

**Comment by rtyler** [comment]
Overall the code looks good to me, but I don't think it has sufficient test coverage to be merged in just yet.

9 times out of 10, I've broken things accidentally in this project that weren't sufficiently tested and made somebody angry. It seems to be more for my benefit than anything :-P

**Comment by mkristian** [comment]
I will give it another round and add more tests. but the test coverage is probably better then you think - it caught a few edge cases for this PR already ;)

File: jruby-gradle-base-plugin/src/integTest/groovy/com/github/jrubygradle/JRubyExecIntegrationSpec.groovy (hunk 3 above)
Comment (mkristian) [inline jruby-gradle-base-plugin/src/integTest/groovy/com/github/jrubygradle/JRubyExecIntegrationSpec.groovy]: the assert I copied from another test which just makes sure the current default version is different from the newVersion

but the test will fail in this case - so I better omit it for "clarity"
